common/client/head_poller_test.go

```diff
@@ -0,0 +1,78 @@
+package client
+
+import (
+	"errors"
+	"math/big"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_Poller(t *testing.T) {
+	// Mock polling function that returns a new value every time it's called
+	var pollNumber int
+	pollFunc := func() (Head, error) {
+		pollNumber++
+		h := head{
+			BlockNumber:     int64(pollNumber),
+			BlockDifficulty: big.NewInt(int64(pollNumber)),
+		}
+		return h.ToMockHead(t), nil
+	}
+
+	// data channel to receive updates from the poller
+	channel := make(chan Head, 1)
+
+	// Create poller and subscribe to receive data
+	poller := NewHeadPoller[Head](time.Millisecond, pollFunc, channel)
+
+	require.NoError(t, poller.Start())
+	defer poller.Unsubscribe()
+
+	// Create goroutine to receive updates from the poller
+	pollCount := 0
+	pollMax := 50
+	go func() {
+		for ; pollCount < pollMax; pollCount++ {
+			h := <-channel
+			assert.Equal(t, int64(pollNumber), h.BlockNumber())
+		}
+	}()
+
+	// Wait for a short duration to allow for some polling iterations
+	time.Sleep(100 * time.Millisecond)
```

Comment:
This makes the test flaky. We should avoid `time.Sleep` in most cases, as we can not be sure how much CPU time we'll actually get. It might take 100ms or 100 seconds for the cycle to complete.

Reply:
Good call, I'll use a channel to signal when the test is done instead.